File: exercises/day13-Keyboard-First-Refactoring/csharp/SleighTripService.Tests/SleighTripServiceShould.cs
using FluentAssertions;
using SleighTripService.Exception;
using Xunit;

namespace SleighTripService.Tests;

public class SleighTripServiceShould
{
    [Fact]
    public void ThrowExceptionWhenNoElfIsLoggedIn()
    {
        // given
        var tripService = new TestableSleighTripService(null);
        var targetElf = new Elf.Elf();

        // when / then
        var act = () => tripService.GetTripsByUser(targetElf);
        act.Should().Throw<ElfNotLoggedInException>();
    }

    //TODO: Please finish testing the GetTripsByUser method!!

    private class TestableSleighTripService : Trip.SleighTripService
    {
        private readonly Elf.Elf? _loggedUser;

        public TestableSleighTripService(Elf.Elf? loggedUser)
        {
            _loggedUser = loggedUser;
        }

        protected override Elf.Elf? GetLoggedUser() => _loggedUser;
    }
}

File: exercises/day13-Keyboard-First-Refactoring/scala/src/test/scala/sleightrip/trip/SleighTripServiceShould.scala
package sleightrip.trip

import sleightrip.exception.ElfNotLoggedInException
import sleightrip.elf.Elf
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class SleighTripServiceShould extends AnyFlatSpec with Matchers {

  "SleighTripService" should "throw exception when no elf is logged in" in {
    // given
    val tripService = new SleighTripService {
      override protected def getLoggedUser: Elf = null // no elf logged in
    }
    val targetElf = new Elf

    // when / then
    assertThrows[ElfNotLoggedInException] {
      tripService.getTripsByUser(targetElf)
    }
  }

  //TODO: Please finish testing the getTripsByUser method!!
}
